apps/rebar/src/rebar_prv_completion.erl

```diff
@@ -0,0 +1,165 @@
+    end.
+
+to_shell("bash") -> bash;
+to_shell("zsh") -> zsh;
+to_shell(Unsupp) ->
+        ?WARN("Unsupported shell found: ~p, default shell will be used.",
+                    [Unsupp]),
+        ?DEF_SHELL.
+
+-spec namespace_to_cmpl_cmds(atom(), [providers:t()]) -> [rebar_completion:cmpl_cmd()].
+namespace_to_cmpl_cmds(default,Providers) ->
+    lists:map(fun(P)->provider_to_cmpl_cmd(P) end,Providers);
+namespace_to_cmpl_cmds(Namespace,Providers) ->
+    Name = atom_to_list(Namespace),
+    [#{name=>Name,
+      commands=>lists:map(fun(P)->provider_to_cmpl_cmd(P) end, Providers),
+      arguments=>[],
+      help=>Name++" namespace"}].
+
+-spec provider_to_cmpl_cmd(providers:t()) -> rebar_completion:cmpl_cmd().
+provider_to_cmpl_cmd(Provider) ->
+    Opts = providers:opts(Provider),
+    Name = providers:impl(Provider),
+    Cmd = getopt_to_cmpl_cmd(atom_to_list(Name),Opts),
+    Help = provider_get(Provider, short_desc),
+    Cmd#{help=>Help}.
+
+-spec getopt_to_cmpl_cmd(string(), [tuple()]) -> rebar_completion:cmpl_cmd().
+getopt_to_cmpl_cmd(Name, Opts) ->
+    Args = [#{short=>S,
+            long=>L,
+            type=>cmpl_arg_type(Spec),
+            help=>H} || {_,S,L,Spec,H} <- Opts],
+    #{name => Name,
+    arguments => Args,
+    commands => [],
+    help => undefined}.
+
+cmpl_arg_type({Type,_Default}) ->
+    Type;
+cmpl_arg_type(Type) ->
+    Type.
+
+%% ad-hoc injection of data for some known providers!
+-spec oracle(rebar_completion:cmpl_cmd(),
+             rebar_completion:cmpl_opts(),
+             rebar_state:t()) -> rebar_completion:cmpl_cmd().
+oracle(#{name:="as"}=Cmd, _CmplOpts, State) ->
+    %% profile completion
+    ConfigProfiles = rebar_opts:get(rebar_state:opts(State), profiles, []),
+    Args = [#{short=>undefined,
+            long=>atom_to_list(ProfileName),
+            help=>undefined,
+            type=>string} || {ProfileName,_} <- ConfigProfiles],
+    Cmd#{arguments=>Args};
+oracle(Cmd,_,_) ->
+    Cmd.
+
+-spec write_completion(iolist(), rebar_state:t(), rebar_completion:cmpl_opts()) -> ok.
+write_completion(CompletionStr, State, #{shell:=Shell, file:=Filename}) ->
+    BaseDir = rebar_dir:base_dir(State),
+    Dest = filename:join(BaseDir, Filename),
+    case filelib:ensure_dir(Dest) of
+        ok ->
+            ?DIAGNOSTIC("Writing completion file for ~p shell to: ~p~n",
+                        [Shell, Dest]),
+            case file:write_file(Dest, CompletionStr, [write, raw]) of
+                ok ->
+                    ok;
+                {error,Err} ->
+                    throw(?PRV_ERROR({error_writing_file,Dest,Err}))
+            end;
+        {error,Err} ->
+            throw(?PRV_ERROR({error_creating_dir,filename:dirname(Dest),Err}))
+    end.
+
+
+%% for some reason providers don't expose some of their attributes via API
+provider_get(P, bare) ->
+    element(5, P);
+provider_get(P, short_desc) ->
+    element(8, P);
+provider_get(P, namespace) ->
+    element(12, P).
```

Comment:
This is a messy bit, which I think you know. Not a blocker for now but we may need to go back at some point and update the providers lib to expose these.

Reply:
I know, if we're able to merge it and bump `providers` version in rebar then I'd do a PR to expose these. @tsloughter ?